cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/ValueNumbering.qll

```diff
@@ -221,6 +235,16 @@ private predicate inheritanceConversionValueNumber(
   valueNumber(instr.getUnary()) = operand
 }
 
+private predicate congruentCopyInstructionTotalValueNumber(
+  CongruentCopyInstructionTotal instr, IRFunction irFunc, IRType type, ValueNumber memOperand,
+  ValueNumber operand
+) {
+  instr.getEnclosingIRFunction() = irFunc and
+  instr.getResultIRType() = type and
```

Comment:
Are there test cases where the type is required to avoid conflating values? If so, I'd be surprised if those cases can't be tweaked so the type is the same.

Reply:
Off the top of my head:
```cpp
double foo(void* p, bool b) {
  if (b) {
    return *(int*)p;
  }
  else {
    return *(float*)p;
  }
}
```
I'm not sure we'd treat the results of the two casts as congruent today, but we probably should. However, we would not want the results of the two indirect loads to be congruent, even though their address operands were congruent.

Reply:
I included the type mainly because all the other constructors did so. I will take a further look tomorrow and see if it actually makes a difference